### ISSUE TITLE
Redirect URI Mismatch

### ISSUE THREAD

**Issue by tebriel** [Shopify/shipit-engine#792]
Hi! I'm using Github enterprise, and have attempted to set up ShipIt and am having a problem with the authentication.
I'm this is amazon with an ELB (port 443) -> Target Group (443 translates to 80) -> Instance (running rails on port 80)

When I attempt to go to the main page, I get the following URL (which translates to a 404 in rails)
```
https://shipit.xxxx.com/auth/failure?message=redirect_uri_mismatch&origin=https%3A%2F%2Fshipit.xxxx.com%2F&strategy=github
```

When I do a TCPDump on the EC2 instance, I see the following:
```
        HTTP/1.1 302 Found
        Location: https://github.xxxx.net/login/oauth/authorize?client_id=MY_CUSTOM_ID&redirect_uri=https%3A%2F%2Fshipit.xxxx.com%2Fgithub%2Fauth%2Fgithub%2Fcallback%3Forigin%3Dhttps%253A%252F%252Fshipit.xxxx.com%252F&response_type=code&scope=email%2Crepo_deployment&state=114a72085634e402687c92367a5ab431dd09261ae3c3422b
        Cache-Control: no-cache
        Set-Cookie: _shipit_session=VERY_LARGE_COOKIE; path=/; secure; HttpOnly
        X-Request-Id: 0088f92e-9e8b-4f92-b4d8-7b00ad3eec88
        X-Runtime: 0.002006
        Strict-Transport-Security: max-age=15552000; includeSubDomains
        Content-Length: 348

        Redirecting to https://github.xxxx.net/login/oauth/authorize?client_id=MY_CUSTOM_ID&redirect_uri=https%3A%2F%2Fshipit.xxxx.com%2Fgithub%2Fauth%2Fgithub%2Fcallback%3Forigin%3Dhttps%253A%252F%252Fshipit.xxxx.com%252F&response_type=code&scope=email%2Crepo_deployment&state=114a72085634e402687c92367a5ab431dd0926[!http]
```

My Github App is configured with User Authorization Callback URL as:
```
https://shipit.xxxx.com/github/auth/github/callback
```

Out of an abundance of caution, I'm also running `force_ssl = True` (since that was hinted at in #684).

At this point I'm at a loss of what I'm doing wrong, when I decode the TCPDump the RedirectURI matches that of what's in the github app.

Thanks for your help.

**Comment by byroot** [comment]
It's weird. `failure?message=redirect_uri_mismatch` suggest the callback URL on the GitHub side doesn't match. But you stated it's good so it must be something else.

Is the "Homepage URL" set to the same domain as well ? IIRC it can cause URI mismatch issues as well.

**Comment by tebriel** [comment]
@byroot yes, homepage url is: `https://shipit.xxxx.com/`

**Comment by tebriel** [comment]
Also I am seeing payloads get delivered (via tcpdump) to the shipit app (registered at `https://shipit.xxxx.com/webhooks`)

**Comment by tebriel** [comment]
<img width="887" alt="screen shot 2018-05-16 at 1 57 13 pm" src="https://user-images.githubusercontent.com/821688/40134844-9ad15eea-5911-11e8-8416-247043d8eb3c.png">
Sorry, figuring out how/then scrubbing took a minute

**Comment by byroot** [comment]
No problem. 

I'm sorry but I can't spot anything wrong here, GitHub returning `redirect_uri_mismatch` doesn't make any sense if the URIs match: https://developer.github.com/apps/managing-oauth-apps/troubleshooting-oauth-app-access-token-request-errors/#redirect-uri-mismatch

I'm afraid you'll have to reach to GitHub support, maybe there is some logs in GitHub Enterprise that could help you better understand why GitHub consider the URL doesn't match ?

**Comment by tebriel** [comment]
Ok, thanks. I'll close out this issue and reach out to them, will add to the bottom of this issue if I find anything useful.

**Comment by tebriel** [comment]
Hey! I just heard back from Github Enterprise support:

```
Thank you for the support bundle, from reviewing the support bundle we can see that the redirect_uri appears to have a parameter appended (origin), which is what is causing the mismatch:

"redirect_uri"=>"https://shipit.xxxx.com/github/auth/github/callback?origin=https%3A%2F%2Fxxxx.pindrop.com%2F"
As per our documentation, additional parameters are not supported:

redirect_uri: The URL in your application where users will be sent after authorization. This must be an exact match to the URL you provided in the User authorization callback URL field when setting up your GitHub App and can't contain any additional parameters.

I located a PR, which was merged and subsequently released last year for the open source OmniAuth GitHub strategy which appears to resolve this issue - https://github.com/omniauth/omniauth-github/pull/73. The OmniAuth GitHub strategy is what Shipit is leveraging to handle the OAuth flow, though it seems that Shipit has not been updated as yet to require a OmniAuth GitHub strategy release containing this fix.

I'd recommend re-opening the issue at https://github.com/Shopify/shipit-engine/issues/792, and seeing if the project maintainers can update the Gemfile in Shipit to use the latest version of the omniauth-github Gem.

Please let us know how you go.

Cheers,
Steven
```

I can take a crack at it, but my rails is pretty shaky, I'll try just updating the gem and seeing if it's a simple upgrade or if there are breaking changes. At least we now know why the `redirect_uri` matches but also doesn't.

**Comment by byroot** [comment]
Interesting. When I debugged your issue initially, I noticed our installation doesn't add the `?origin=` param. 

I'll try to figure out where it comes from tomorrow (today is an holiday). In the meantime, could you tell me which version of `omniauth` & `omniauth-github` your installation use ?

```
$ bundle show omniauth
$ bundle show omniauth-github
```

We set relatively strict version requirements, but I prefer to be sure I look at the proper version of the dependency.

**Comment by tebriel** [comment]
I cloned the `shipit-engine` source (which has 1.1.2 defined here: https://github.com/Shopify/shipit-engine/blob/master/shipit-engine.gemspec#L31) and modified it to be 1.3.0, pointed bundler at the cloned folder, and the auth flow works now (just dealing with asset pipeline issues because I haven't deployed rails in forever). Here's the Gemfile.lock generated by the app before I started installing the gem from the source: https://gist.github.com/tebriel/03e196f8f0f305c947722f80111e8c3f. Note it points at 1.1.2 as well.

**Comment by tebriel** [comment]
After bumping that gemspec to use 1.3.0 and then fixing my asset pipeline issues (so much rails knowledge is flooding back right now) it looks like i have a working system! Will try to set up a deploy later this afternoon. The updated gem doesn't seem to cause any immediate harm.

**Comment by byroot** [comment]
Looking at `omniauth-github` history, it was indeed fixed in `1.2.3` https://github.com/omniauth/omniauth-github/commit/d8d5c217f560b7021760882c3c7bdc3e971e0fd8

I'll update the dependency and release a new version of `shipit-engine`.

Thanks a lot for figuring this out.

**Comment by casperisfine** [comment]
Released as `v0.22.1`.

**Comment by tebriel** [comment]
Thanks so much for your help! I’ll upgrade after I get into the office.

On May 22, 2018, at 4:18 AM, Jean byroot Boussier <notifications@github.com<mailto:notifications@github.com>> wrote:


Released as v0.22.1.
